### ISSUE TITLE
PFS has to be aware of `Settle_timeouts`

### ISSUE THREAD

**Issue by err508** [raiden-network/raiden-services#5]
Currently, the path finding is performed with the constraint of available `channel capacity`. Additionally the pfs has to - given a `reveal_timeout` of a payment to be routed - incorporate the `settle timeout` of the hops as a constraint. Else, the mediating nodes will reject forwarding the payment. The PFS may learn the `settle timeout` from the onchain channel opening events. 

### To Do
- [x] Store settle timeout for each channel
- [x] ~~Add reveal timeout to the balance update messages from the Raiden client~~(reveal_timeout will be added to the Balance Proof that is sent to the PFS)
- [x] Incorporate the above information in the routing algorithm
- [x] ~~Return the best lock expiration setting for each returned path~~ (decided not to)

**Comment by Dominik1999** [comment]
@hackaugusto, @LefterisJP, @karlb, @palango open for discussion:

We now included `settle_timeout` and `reveal_timeout` into the PFS. There are three open questions to discuss:

1. Should every client for every channel publish its `reveal_timeout` together with its `Balance Proof` or only after a certain time interval, e.g. every 10 Minutes?

2. Should the client's route request be customizable with respect to a certain ratio `(:= c)` over all hops of `settle_timeout / reveal_timeout`, therefore say "I only want to see routes with `c > 2`"? (At the moment it is 2 for the clients)

3. Should the PFS optimize routes with respect (a) to the highest ratio `settle_timeout / reveal_timeout` or (b) respond to the payment initiator with a suggestion for `c` for the respective route?

Could you add your thoughts, such that we come to a decision until 30.01.2019?
---

**Summary of the problem:** 

No mediator would forward a payment if `lock_expiration (= c * reveal_timeout) < settle_timeout` whereas `c` is a constant currently at `2` and `settle_timeout` is the `settle_timeout` of the channel the mediator receives the payment through.

This condition is already checked in the client.

**Example:** In case mediator B would forward a payment and not follow the above condition
```
A payment with reveal timeout of 50 blocks is send on this route:

   10       100
A----> B ----> C,

where the numbers over the arrows depict the settle timeouts of the respective channel
```
After the payment is sent, `A` settles the channel, unlocks and reveals the secret. `C` can now wait `40` Blocks to unlock the payment. Then, `B` can not use the secret in the `(A,B)` channel anymore, as the channel is already settled. B loses tokens.

**Comment by hackaugusto** [comment]
> Should every client for every channel publish its reveal_timeout together with its Balance Proof or only after a certain time interval, e.g. every 10 Minutes?

I don't see why. To do this we would need:

- A new message, to update only the reveal timeout
- Raiden: A new timer and client methods, to update on a separate interval
- PFS: A new endpoint, just for this

To very little gain. The best I can think is to "optimize" the size of the updates with the balance proof. But the reveal timeout, for a good packed representation, will be a single byte, and for a JSON it will around 25 bytes. I don't think the savings in space are worth it. 

> Should the client's route request be customizable with respect to a certain ratio (:= c) over all hops of settle_timeout / reveal_timeout, therefore say "I only want to see routes with c > 2"? (At the moment it is 2 for the clients)

Is `c` the constant in the code snippet mentioned by paul? `self.raiden.config['reveal_timeout'] * 2:` I would not worry about this in the PFS at all, it's just a check for a "good" reveal timeout value.

What the PFS can do is the following:

- For every channel is a path:
  1. Get the minimum settle timeout
  2. Get the maximum reveal timeout
  3. Discard the route if reveal timeout >= settle timeout

If condition 3 above is ever met, it means at some point the transfer will get stuck, because a node will consider it unsafe to mediate the transfer. (Note: The above is simplified since it's not taking care of the order, but I think this is fine)

**Comment by palango** [comment]
> Is c the constant in the code snippet mentioned by paul?

c is the ration of *settle_timeout*/*reveal_timeout*.

As you wrote above:
- `c <= 1` is unsafe and should be discarded
- `c close to 1` assumes fast payments and might force nodes to go on-chain fast
- `c = 2` is the hardcoded default in raiden.

So you propose to just filter out routes for `c <=1`, but we think it might be nice to make this additional information we have usable for the nodes, as it might be not in a nodes interest to do a payment which has a transfer with `c= 1.01` in the path.

**Comment by karlb** [comment]
>     1. Should every client for every channel publish its `reveal_timeout` together with its `Balance Proof` or only after a certain time interval, e.g. every 10 Minutes?

I don't see any harm in sending it along every time.

>     2. Should the client's route request be customizable with respect to a certain ratio `(:= c)` over all hops of `settle_timeout / reveal_timeout`, therefore say "I only want to see routes with `c > 2`"? (At the moment it is 2 for the clients)

I don't see much value in having it customizable by the client. Why would the initiator want a much different value than "slightly above the default threshold required for forwarding"?
 
>     3. Should the PFS optimize routes with respect (a) to the highest ratio `settle_timeout / reveal_timeout` or (b) respond to the payment initiator with a suggestion for `c` for the respective route?

The PFS should not generate routes which are sure to fail and penalize routes which are likely to fail due to this constraint. So maybe something like (assuming nodes running the default client will stop forwarding at a  c=2 threshold)
c < 2: delete route
c < 2.5 penalize route
c >= 2.5 no penalty

If nodes want to choose vastly different thresholds in the future, this would get more complicated. But I don't see why such flexibility would be needed.

**Comment by LefterisJP** [comment]
> Should every client for every channel publish its reveal_timeout together with its Balance Proof or only after a certain time interval, e.g. every 10 Minutes?

I also don't see a problem with this one. Just include it with the BP.

> Should the client's route request be customizable with respect to a certain ratio (:= c) over all hops of settle_timeout / reveal_timeout, therefore say "I only want to see routes with c > 2"? (At the moment it is 2 for the clients)

I think we should go with the simple approach of simply `c<2` (assuming c == 2 is the client default) delete route and accept all others.

I don't see why the routes between 2 and 2.5 would need to be penalized. `c==2` is quite safe already. With `c==2` I would start thinking only of the capacities and not of timeouts.

**Comment by karlb** [comment]
> I don't see why the routes between 2 and 2.5 would need to be penalized. c==2 is quite safe already. With c==2 I would start thinking only of the capacities and not of timeouts.

I thought we need some buffer to accommodate the time that passes between the route planning and the mediator checking the timeouts. But since we are talking about time intervals and not about points in time, that is actually not necessary. So I agree that just using `c==2` as a fixed threshold is a good choice.,

**Comment by Dominik1999** [comment]
Thanks for all your points! We now just
- skip all routes with c < 2 (#46) and 
- add the reveal_timeout to every balance proof update sent to the PFS (to be defined in an issue)

For me, this was a productive, efficient decision making process

**Comment by Dominik1999** [comment]
closed by #39
